cdi/src/main/java/io/smallrye/config/inject/ConfigExtension.java

```diff
@@ -159,36 +162,50 @@ protected void validate(@Observes AfterDeploymentValidation adv) {
             if ((!configNames.contains(name) && ConfigProducerUtil.getRawValue(name, config) == null)
                     && !isIndexed(type, name, config)) {
                 if (configProperty.defaultValue().equals(ConfigProperty.UNCONFIGURED_VALUE)) {
-                    adv.addDeploymentProblem(InjectionMessages.msg.noConfigValue(name));
+                    adv.addDeploymentProblem(
+                            InjectionMessages.msg.noConfigValue(name, formatInjectionPoint(injectionPoint)));
                     continue;
                 }
             }
 
             try {
                 // Check if the value can be injected. This may cause duplicated config reads (to validate and to inject).
                 ConfigProducerUtil.getValue(injectionPoint, config);
-            } catch (IllegalArgumentException e) {
-                adv.addDeploymentProblem(InjectionMessages.msg.illegalConversion(name, type));
-            } catch (NoSuchElementException e) {
-                adv.addDeploymentProblem(e);
+            } catch (Exception e) {
+                adv.addDeploymentProblem(InjectionMessages.msg.retrieveConfigFailure(name, formatInjectionPoint(injectionPoint),
+                        e.getLocalizedMessage(), e));
             }
         }
 
-        Set<ConfigMappings.ConfigMappingWithPrefix> configMappingsWithPrefix = new HashSet<>();
         for (AnnotatedType<?> configMapping : configMappings) {
-            configMappingsWithPrefix
-                    .add(configMappingWithPrefix(configMapping.getJavaClass(), getPrefixFromType(configMapping)));
+            registerAndValidateConfigMapping(
+                    configMappingWithPrefix(configMapping.getJavaClass(), getPrefixFromType(configMapping)), adv,
+                    (SmallRyeConfig) config);
         }
 
         for (InjectionPoint injectionPoint : configMappingInjectionPoints) {
-            getPrefixFromInjectionPoint(injectionPoint).ifPresent(prefix -> configMappingsWithPrefix
-                    .add(configMappingWithPrefix((Class<?>) injectionPoint.getType(), prefix)));
+            getPrefixFromInjectionPoint(injectionPoint).ifPresent(prefix -> registerAndValidateConfigMapping(
+                    configMappingWithPrefix((Class<?>) injectionPoint.getType(), prefix), adv, (SmallRyeConfig) config));
         }
 
+    }
+
+    private void registerAndValidateConfigMapping(ConfigMappings.ConfigMappingWithPrefix cm, AfterDeploymentValidation adv,
+            SmallRyeConfig config) {
         try {
-            ConfigMappings.registerConfigMappings((SmallRyeConfig) config, configMappingsWithPrefix);
+            ConfigMappings.registerConfigMappings(config, Collections.singleton(cm));
```

Comment:
Aren't we creating multiple instances of `ConfigMappingProvider`? I'm not sure if this is ok. For example, when validating properties, each instance will report unused properties because they are only looking into their mapping subset and have no idea if unused properties are being mapped in another instance of `ConfigMappingProvider`.

Reply:
Yes, that would be the case. I'm not sure exactly what you mean though? The point of commit https://github.com/smallrye/smallrye-config/pull/513/commits/65d869887be15ef447fd2a1dfc69421d60b22065 was to deal with this fact. IIRC correctly, the `ConfigMappingProvider` just gets used to populate `this.mappings` in `ConfigMappings` via the `registerConfigMappings()` method which was updated:
https://github.com/smallrye/smallrye-config/blob/7a0e7685ae67e3e4317602573ed5f4884ccdc531/implementation/src/main/java/io/smallrye/config/ConfigMappings.java#L26-L35

Also, isn't this what you suggested in this comment? https://github.com/smallrye/smallrye-config/pull/513#discussion_r603264073

Have you got an example test that this implementation may break?

Reply:
Did you see that `registerAndValidateConfigMapping()` is being called as each `configMapping` and `injectionPoint` is being discovered? As opposed to before when they were all being collected, then registered. I think this may have been the resolve to the "unused properties being mapped" issue you mentioned:  https://github.com/smallrye/smallrye-config/blob/7a0e7685ae67e3e4317602573ed5f4884ccdc531/cdi/src/main/java/io/smallrye/config/inject/ConfigExtension.java#L180-L189

Reply:
@radcortez I think it would be useful to have an example of what you think wouldn't work. At the moment Joe's code is passing all of the existing tests.

Reply:
Ok, let me try to implement a test.

Reply:
Please check: https://github.com/smallrye/smallrye-config/pull/552. I think we should make these tests more self contained.

If I apply this change plus the current registration proposal, we get a validation error, because some properties will be reported as not being used (they are, but the check is performed in different instances). Please check https://github.com/radcortez/smallrye-config/commits/multiple-config-mapping-provider and run `ConfigMappingInjectionTest`.

Reply:
I agree, #552 is a very good change, and makes the tests a lot more clear. Note that with this change, this Javadoc would need updating:
https://github.com/smallrye/smallrye-config/blob/7a0e7685ae67e3e4317602573ed5f4884ccdc531/cdi/src/test/java/io/smallrye/config/inject/ValidateInjectionTest.java#L45-L47

Isn't the error because you're missing this important change?
https://github.com/smallrye/smallrye-config/blob/7a0e7685ae67e3e4317602573ed5f4884ccdc531/implementation/src/main/java/io/smallrye/config/ConfigMappings.java#L26-L35

Reply:
I've rebased with main locally, and with the new Exception messages in this PR I'm getting
```
org.jboss.weld.exceptions.DeploymentException: SRCFG02004: Failed to create @ConfigMapping bean with prefix client for interface io.smallrye.config.inject.ConfigMappingInjectionTest$Server. Configuration validation failed:
java.util.NoSuchElementException: SRCFG00014: The config property client.port is required but it could not be found in any config source
```
for this line which sounds about right to me?
https://github.com/smallrye/smallrye-config/blob/d7d2b367db5db2125d1d832ff46b80315782f4d3/cdi/src/test/java/io/smallrye/config/inject/ConfigMappingInjectionTest.java#L102

Reply:
Can you push to the PR, even if it is failing so I can have a better look? Thanks!

Reply:
I'll push what I had, although I have no idea how this is going to work with the merging of https://github.com/smallrye/smallrye-config/pull/561.

The branch currently fails tests with `does not map to any root` Exception messages. I commented out the properties throwing these Exceptions to achieve the Exception messages in https://github.com/smallrye/smallrye-config/pull/513#discussion_r623205463. 

I'm going to try to create a separate PR which adds the improved Injection Exception messages from this PR without touching the Config Mapping and Config Properties code, so that we at least have some improvement to the messages.

Reply:
> I'm going to try to create a separate PR

https://github.com/smallrye/smallrye-config/pull/562

@radcortez Am I correct in thinking that if a ConfigMapping interface is injected, with lets say `prefix="client"`, then SmallRye Config "validates" every config property that exists in all Config Sources, for the given prefix, and if there exists a config property which isn't used by any ConfigMapping interface, a `<config property> does not map to any root Exception` is thrown? If so, why? And where can I find documentation on this please?

Reply:
Not every config property, but configs in the namespace of the ConfigMapping. The ideia is that you get a validation in case you forgot to map a property in your mapping class.

Reply:
Is [this documentation](https://smallrye.io/docs/smallrye-config/mapping/mapping.html) the only doc for ConfigMapping? (and of course [the Javadoc](https://github.com/smallrye/smallrye-config/blob/main/implementation/src/main/java/io/smallrye/config/ConfigMapping.java))